src/npfctl/npfctl.c

```diff
@@ -275,6 +278,96 @@ npfctl_print_addrmask(int alen, const char *fmt, const npf_addr_t *addr,
 	return buf;
 }
 
+static int
+npfctl_table_type(const char *typename)
+{
+	int i, type = 0;
+
+	static const struct tbltype_s {
+		const char *name;
+		u_int type;
+	} tbltypes[] = {
+		{ "ipset",	NPF_TABLE_IPSET	},
+		{ "lpm",	NPF_TABLE_LPM	},
+		{ "const",	NPF_TABLE_CONST	},
+		{ NULL,		0		}
+	};
+
+	for (i = 0; tbltypes[i].name != NULL; i++) {
+		if (strcmp(typename, tbltypes[i].name) != 0) {
+			continue;
+		}
+		type = tbltypes[i].type;
+		break;
+	}
+
+	return type;
+}
+
+__dead static void
+npfctl_table_replace(int fd, int argc, char **argv)
+{
+	const char *name, *newname, *fname;
+	nl_config_t *ncf;
+	nl_table_t *t;
+	npf_error_t errinfo;
+	int c, tid;
+	u_int type = 0;
+
+	name = newname = argv[0];
+	optind = 2;
+	while ((c = getopt(argc, argv, "n:t:")) != -1) {
+		switch (c) {
+		case 't':
+			type = npfctl_table_type(optarg);
+			if (!type) {
+				errx(EXIT_FAILURE, "unsupported table type '%s'", optarg);
+			}
+			break;
+		case 'n':
+			newname = optarg;
+			break;
+		default:
+			fprintf(stderr,
+			    "Usage: %s table <tid> replace [-n <newid>] [-t ipset|lpm|const] <table-file>\n",
+			    getprogname());
+			exit(EXIT_FAILURE);
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	if (argc != 1) {
+		usage();
+	}
+
+	fname = argv[0];
+
+	/* Get existing config to lookup ID of existing table */
+	if (!(ncf = npf_config_retrieve(fd))) {
+		err(EXIT_FAILURE, "npf_config_retrieve()");
+	}
+	if (!(t = npfctl_table_getbyname(ncf, name))) {
+		errx(EXIT_FAILURE, "no existing table '%s' to replace", name);
+	}
+	tid = npf_table_getid(t);
+	if (!type) {
+		type = npf_table_gettype(t);
+	}
+	npf_config_destroy(ncf);
+
+	/* Build a new table */
+	npfctl_build_table(newname, tid, type, fname);
```

Comment:
It would be nice to support `stdin` as input, if the file name is specified as `-`.

Reply:
Agreed, I'll look at doing that.

Splitting `npfctl_build_table()` (as mentioned in the comment above) would also help achieve this by allowing a `FILE *` to be passed rather than filename. This would allow the command processor to check for "-" and pass `stdin`. Sound OK?